File: examples/calculator-java-hexagonal-cqrs/src/main/java/com/oopforge/example/calculator/adapter/persistence/InMemoryCalculationRepository.java
package com.oopforge.example.calculator.adapter.persistence;

import com.oopforge.example.calculator.application.command.CalculationRepository;
import com.oopforge.example.calculator.application.query.HistorySummary;
import com.oopforge.example.calculator.domain.Calculation;

/**
 * Write adapter: persists the aggregate and projects the read model.
 */
public class InMemoryCalculationRepository implements CalculationRepository {

    private final CalculationStore store;

    public InMemoryCalculationRepository(CalculationStore store) {
        this.store = store;
    }

    @Override
    public void save(Calculation calculation) {
        store.putCalculation(calculation);
        store.appendHistory(HistorySummary.from(calculation));
    }
}

File: examples/calculator-python-hexagonal-cqrs/app/infrastructure/repositories/in_memory_calculation_repository.py
from app.application.query.history_summary import HistorySummary
from app.domain.calculation.model import Calculation
from app.infrastructure.repositories.store import CalculationStore


class InMemoryCalculationRepository:
    """Write adapter: persists the aggregate and projects the read model."""

    def __init__(self, store: CalculationStore) -> None:
        self._store = store

    def save(self, calculation: Calculation) -> None:
        self._store.calculations[calculation.id] = calculation
        self._store.history.append(HistorySummary.from_calculation(calculation))
